File: array/Solution18.java
import java.util.List;
import java.util.LinkedList;
import java.util.Arrays;

class Solution18 {
    public List<List<Integer>> fourSum(int[] nums, int target) {
    	Arrays.sort(nums);
    	List<List<Integer>> result = new LinkedList<>();
    	for(int point1 = 0; point1 < nums.length; point1++){
    		if(point1 > 0 && nums[point1] == nums[point1 - 1]){
    			continue;
    		}
    		for(int point2 = point1 + 1; point2 < nums.length; point2++){
    			if(point2 != point1 + 1 && nums[point2] == nums[point2 - 1]){
    				continue;
    			}

    			int tar = target -(nums[point1] + nums[point2]);
    			for(int point3 = point2 + 1; point3 < nums.length; point3++){
    				if(point3 != point2 + 1 && nums[point3] == nums[point3 - 1]){
    					continue;
    				}

    				int point4 = nums.length - 1;
    				while(point4 > point3 && nums[point3] + nums[point4] > tar){
    					point4 -= 1;
    				}
    				if(point3 == point4){
    					continue;
    				}

    				if(nums[point3] + nums[point4] == tar){
    					List<Integer> cur = new LinkedList<>();
    					cur.add(nums[point1]);
    					cur.add(nums[point2]);
    					cur.add(nums[point3]);
    					cur.add(nums[point4]);
    					result.add(cur);
    					continue;
    				}
    			}
    		}
    	}
    	return result;
    }

    public static void main(String[] args){
    	Solution18 s = new Solution18();
    	int[] test1 = new int[]{1, 0, -1, 0, -2, 2};
    	List<List<Integer>> result = s.fourSum(test1, 0);
    	
    }
}

File: C++/Solution18.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution18 {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> result;
        int size = nums.size();
        sort(nums.begin(), nums.end());
        for(int index1 = 0; index1 < size; index1++){
            if(index1 > 0 && nums[index1] == nums[index1 - 1]){
                continue;
            }
            for(int index2 = index1 + 1; index2 < size; index2++){
                if(index2 > index1 + 1 && nums[index2] == nums[index2 - 1]){
                    continue;
                }

                int tar = target - nums[index1] - nums[index2];
                int index3 = index2 + 1, index4 = size - 1;
                while(index4 > index3){
                    int sum = nums[index3] + nums[index4];
                    if(sum == tar){
                        result.push_back({nums[index1], nums[index2], nums[index3], nums[index4]});
                        index3++;
                        index4--;
                        while(index3 < size && nums[index3] == nums[index3 - 1]){
                            index3++;
                        }
                        while(index4 >= 0 && nums[index4] == nums[index4 + 1]){
                            index4--;
                        }
                    }
                    else if(sum > tar){
                        index4--;
                        while(index4 >= 0 && nums[index4] == nums[index4 + 1]){
                            index4--;
                        }
                    }
                    else{
                        index3++;
                        while(index3 < size && nums[index3] == nums[index3 - 1]){
                            index3++;
                        }
                    }
                }
            }
        }
        return result;
    }
};